crates/aptos-crypto/src/poseidon_bn254/mod.rs

```diff
@@ -0,0 +1,147 @@
+
+/// The maximum number of input scalars that can be hashed using the Poseidon-BN254 hash function
+/// exposed in `hash_scalars`.
+pub const MAX_NUM_INPUT_SCALARS: usize = 16;
+
+/// Macro for Poseidon-BN254-hashing a vector of scalars.
+macro_rules! neptune_hash {
+    ($elems:expr, $constants:expr) => {{
+        let mut hasher = Poseidon::new(&$constants);
+        hasher.reset();
+        for elem in $elems.into_iter() {
+            hasher.input(elem.into()).expect("Too many inputs");
+        }
+        hasher.hash_in_mode(OptimizedStatic);
+        hasher.elements[0]
+    }};
+}
+
+/// Given an array of up to `MAX_NUM_INPUT_SCALARS` field elements (in the BN254 scalar field), hashes
+/// them using Poseidon-BN254 into a single field element.
+pub fn hash_scalars(inputs: Vec<ark_bn254::Fr>) -> anyhow::Result<ark_bn254::Fr> {
+    if inputs.is_empty() || inputs.len() > MAX_NUM_INPUT_SCALARS {
+        bail!(
+            "Poseidon-BN254 needs > 0 and <= {} inputs, but was called with {} inputs",
+            MAX_NUM_INPUT_SCALARS,
+            inputs.len()
+        );
+    }
+
+    let result = match inputs.len() {
+        1 => neptune_hash!(inputs, POSEIDON_1),
+        2 => neptune_hash!(inputs, POSEIDON_2),
+        3 => neptune_hash!(inputs, POSEIDON_3),
+        4 => neptune_hash!(inputs, POSEIDON_4),
+        5 => neptune_hash!(inputs, POSEIDON_5),
+        6 => neptune_hash!(inputs, POSEIDON_6),
+        7 => neptune_hash!(inputs, POSEIDON_7),
+        8 => neptune_hash!(inputs, POSEIDON_8),
+        9 => neptune_hash!(inputs, POSEIDON_9),
+        10 => neptune_hash!(inputs, POSEIDON_10),
+        11 => neptune_hash!(inputs, POSEIDON_11),
+        12 => neptune_hash!(inputs, POSEIDON_12),
+        13 => neptune_hash!(inputs, POSEIDON_13),
+        14 => neptune_hash!(inputs, POSEIDON_14),
+        15 => neptune_hash!(inputs, POSEIDON_15),
+        16 => neptune_hash!(inputs, POSEIDON_16),
+        _ => bail!(
+            "Poseidon-BN254 was called with {} inputs, more than the maximum 16 allowed inputs.",
+            inputs.len()
+        ),
+    };
+
+    Ok(result.into())
+}
+
+/// Given an string and `max_bytes`, it pads the byte array of the string with zeros up to size `max_bytes`,
+/// packs it to scalars, and returns the hash of the scalars.
+///
+/// This function calls `pad_and_pack_bytes_to_scalars_no_len` safely as strings will not contain the zero byte except to terminate.
+pub fn pad_and_hash_string(str: &str, max_bytes: usize) -> anyhow::Result<ark_bn254::Fr> {
+    pad_and_hash_bytes_with_len(str.as_bytes(), max_bytes)
+}
+
+/// Packs the bytes to a vector of scalars (see `pack_bytes_to_scalars`) and hashes the scalars via
+/// `hash_scalars`.
+///
+/// Note: The byte packing encodes the length of the bytes properly so as to avoid collisions when
+/// hashing, say, 0x00 versus 0x0000.
+///
+/// WARNING: We do not expose this function to avoid unnecessary bugs, since for SNARK circuits we
```

Comment:
Does this really need a WARNING tag? The comment itself is good

Reply:
🤷